### PR TITLE
Add configuration to hide foreground notifications (Android)

### PR DESCRIPTION
**Pull request by willatrush**
hi @colmharte would appreciate your help again

We are still getting some feedback from Playstore (NZ COVID Tracer App 3.0.2) that the notification stuck issue is still occuring:

> Constantly stuck checking exposure keys on my s10. Chewing through battery 5% in 10 min at this rate my battery will be flat in just over 3 hours. If this can be fixed will need to delete app.

> Still getting the infinite checking keys bug. Apart from that it seems to do what it's supposed to.

It seems that workers with a FGS do not have a timeout, while normal workers do timeout after 10 min

Added a configuration to hide foreground notifications altogether

### CODE REVIEW

File: src/exposure-provider.tsx [hunk 1]
```diff
@@ -197,12 +199,13 @@ export const ExposureProvider: React.FC<ExposureProviderProps> = ({
     async function checkSupportAndStart() {
       await supportsExposureApi();
 
+      await configure();
```
Comment (willatrush) [inline src/exposure-provider.tsx]: Moved this line out of the condition as the job still runs for users with enf off / unsupported.

File: android/src/main/java/ie/gov/tracing/nearby/ProvideDiagnosisKeysWorker.java [hunk 2]
```diff
@@ -80,7 +80,6 @@ public ProvideDiagnosisKeysWorker(@NonNull Context context,
     secureRandom = new SecureRandom();
     repository = new ExposureNotificationRepository(context);
     this.context = context;
-    setForegroundAsync(createForegroundInfo());
```
Comment (ilya-rush) [inline android/src/main/java/ie/gov/tracing/nearby/ProvideDiagnosisKeysWorker.java:83]: For additional context, an engineer from Google recommended that `setForegroundAsync` should only be called inside `startWork` and shouldn't be in the constructor.

**Comment by colmharte** [comment]
@willatrush I have concerns that this change is not fixing your actual problem. Is the battery being consumed by ENS or your covid app? Whether you are using a foreground service or not will not change battery consumption. The 10 minute timeout for non foreground services also is not the solution. I can't see how this module would be spending 10 minutes plus processing. So hiding the foreground notification is masking the issue rather than solving it

File: src/exposure-provider.tsx [hunk 3]
```diff
@@ -161,7 +162,8 @@ export const ExposureProvider: React.FC<ExposureProviderProps> = ({
   callbackNumber = '',
   analyticsOptin = false,
   notificationRepeat = 0,
-  certList = ''
+  certList = '',
+  hideForeground
```
Comment (colmharte) [inline src/exposure-provider.tsx]: default this to false

File: src/exposure-provider.tsx (hunk 1 above)
Comment (colmharte) [reply src/exposure-provider.tsx]: You should not call configure before isReady is true, this will result in invalid config being passed to the module, please remove this change. For example app has not yet loaded users refresh and auth token so calling configure too soon will result in empty tokens being stored in shared prefs. 
The background process checks for valid config when it runs so not calling config is not an issue

File: src/exposure-provider.tsx [hunk 4]
```diff
@@ -197,12 +199,13 @@ export const ExposureProvider: React.FC<ExposureProviderProps> = ({
...
       await supportsExposureApi();
 
+      await configure();
+
       // Start as soon as we're able to
       if (
         isReady &&
         state.permissions.exposure.status === PermissionStatus.Allowed
       ) {
-        await configure();
```
Comment (colmharte) [inline src/exposure-provider.tsx]: As above, please undo this change

**Review by colmharte — CHANGES_REQUESTED** [review]

File: android/src/main/java/ie/gov/tracing/nearby/ProvideDiagnosisKeysWorker.java [hunk 5]
```diff
@@ -129,15 +128,27 @@ private void updateLastRun() {
   @NonNull
   @Override
   public ListenableFuture<Result> startWork() {
+      // validate config set before running
+      final String server = SharedPrefs.getString("serverUrl", this.context);
```
Comment (willatrush) [inline android/src/main/java/ie/gov/tracing/nearby/ProvideDiagnosisKeysWorker.java:132]: moved this condition up so on first run, if config wasn't set, it won't show a foreground notification

File: src/exposure-provider.tsx (hunk 1 above)
Comment (willatrush) [reply src/exposure-provider.tsx]: reverted. instead move the serverUrl config check further up

File: src/exposure-provider.tsx (hunk 3 above)
Comment (willatrush) [reply src/exposure-provider.tsx]: done

**Comment by ilya-rush** [comment]
@colmharte my 2 pence about this change in the light of your comment:

> I have concerns that this change is not fixing your actual problem. Is the battery being consumed by ENS or your covid app? Whether you are using a foreground service or not will not change battery consumption. The 10 minute timeout for non foreground services also is not the solution. I can't see how this module would be spending 10 minutes plus processing. So hiding the foreground notification is masking the issue rather than solving it

After the previous fixes for the stuck "Checking for exposures" notification from the foreground service we haven't been able to reproduce that issue yet, so we are just trusting our users on this one and applying some general thinking. To answer your question about battery consumption, we don't know. We had some battery issues that weren't related to EN or the plugin, but those should be fixed in our app in the version users reported it for again.

While we are looking for a way to reproduce the "stuck" issue, the general thinking is:
- Service being foreground can (somehow) hypothetically be an essential prerequisite for it getting stuck. In this case not running it in foreground could solve the issue (although I admit, this is a stab in the dark).
- Or it's getting stuck somewhere else. In this case not running it in foreground can partially mitigate the problem by enforcing  a 10m timeout so that if it's stuck, it's not stuck forever.

That reasoning alone may not justify modification of the plugin used by many developers enough. However, I think there could be some unrelated functional benefit that's generally useful. For the servers that allow for processing in under 10m (I'm assuming most if not all of them), developers would be able to choose how they want their apps to behave: stick with the defaults when checks are surfaced to the users or make checks invisible.

**Review by colmharte — APPROVED** [review]
